### PR TITLE
Allow configuring Project Manager server host and port

### LINKED ISSUE #6730: [/127.0.0.1:30535] Address in use
I have downloaded https://github.com/enso-org/enso/releases/tag/2023.2.1-nightly.2023.5.17 release and tried to launch it:

```
~$ /enso-linux-2023.2.1-nightly.2023.5.17.AppImage
Electron application is ready.
╭ Starting the application 
│ ╭ Starting the backend process with the following options: . 
│ │ Backend has been spawned (pid = 160705).
│ ╰ Done in 5.4 ms. 
│ ╭ Starting the content server. 
│ │ Server started on port 8081.
│ │ Serving files from '/home/devel/tmp/.mount_enso-l1VDuon/resources/app.asar/assets'.
│ ╰ Done in 13.2 ms. 
│ ╭ Creating the window. 
[info] [2023-05-17T11:39:00.294Z] [org.enso.projectmanager.boot.ProjectManager$] Starting Project Manager...
timestamp=2023-05-17T11:39:00.439621Z level=ERROR thread=#zio-fiber-1 message="" cause="Exception in thread "zio-fiber-5" akka.stream.impl.io.ConnectionSourceStage$$anon$1$$anon$2: Bind failed because of java.net.BindException: [/127.0.0.1:30535] Address in use
        at org.enso.projectmanager.boot.ProjectManager.bindServer(ProjectManager.scala:298)
        at org.enso.projectmanager.boot.ProjectManager.mainProcess(ProjectManager.scala:74)
        at org.enso.projectmanager.boot.ProjectManager.runOpts(ProjectManager.scala:238)
        at org.enso.projectmanager.boot.ProjectManager.runArgs(ProjectManager.scala:122)
        Suppressed: java.net.BindException: [/127.0.0.1:30535] Address in use
                at java.base@11.0.18/sun.nio.ch.Net.bind0(Net.java)
                at java.base@11.0.18/sun.nio.ch.Net.bind(Net.java:459)
                at java.base@11.0.18/sun.nio.ch.Net.bind(Net.java:448)
                at java.base@11.0.18/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:227)
                at java.base@11.0.18/sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:80)
                at akka.io.TcpListener.liftedTree1$1(TcpListener.scala:60)
                at akka.io.TcpListener.<init>(TcpListener.scala:57)
                at java.base@11.0.18/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
                at akka.util.Reflect$.instantiate(Reflect.scala:73)
                at akka.actor.ArgsReflectConstructor.produce(IndirectActorProducer.scala:101)
                at akka.actor.Props.newActor(Props.scala:226)
                at akka.actor.ActorCell.newActor(ActorCell.scala:616)
                at akka.actor.ActorCell.create(ActorCell.scala:643)
                at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:514)
                at akka.actor.ActorCell.systemInvoke(ActorCell.scala:536)
                at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:295)
                at akka.dispatch.Mailbox.run(Mailbox.scala:230)
                at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
                at java.base@11.0.18/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
                at java.base@11.0.18/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
                at java.base@11.0.18/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
                at java.base@11.0.18/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
                at java.base@11.0.18/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

Are we really using the same **fixed port**? What if it is taken? What if one wants to run two Enso IDE on the same computer (by different user)? Shouldn't the port rather be dynamically allocated?

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Closes #6730
  - Changes config to allow environment variables to override server host and port
  - Adds port scanning to Electron app to ensure the PM is started at a free port

### Important Notes
- `SERVER_PORT=abcd enso.AppImage` does NOT work. It would not be difficult to implement, but it probably needs discussion on how exactly it should be implemented - for example, `SERVER_PORT` is quite a generic name, should the Electron app pass though something like `ENSO_PM_SERVER_PORT` to the PM as `SERVER_PORT` instead?

:warning: Port scanning is *only* implemented in the JS frontend. It is not implemented:
- In Scala, because the JS/Rust code calling it needs to know the port as well. There shouldn't be any problems with adding port scanning though, if that's desired
- In Rust, because I'm not sure parsing the host and port from a string is a good idea.
  - (This also applies to JS, but it *must* work in JS, and port scanning is already a dependency there so it's quite a bit easier)

- QA *will* need a new PM (`sbt buildProjectManagerDistribution` or `./run backend sbt` -> `buildProjectManagerDistribution`), and the path must be supplied as: `-engine.project-manager-path=path/to/new/pm/here`

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by somebody1234** [comment]
i don't know whether this needs tests on the scala side

**Comment by somebody1234** [comment]
So it seems like running two instances side-by-side by the *same* user should be unsupported - the Project Managers seem to want to acquire locks on the same files for compilation. So, when testing, instead a dummy HTTP server should be run on port 30535

**Comment by xvcgreg** [comment]
@PabloBuchu what's the status of this task?

**Comment by somebody1234** [comment]
closing until this becomes an issue again

**Comment by somebody1234** [comment]
i've done some basic testing only - by using `ide watch` with a http server already running at port `30535` (which makes it start the PM at port 30536):
- :heavy_check_mark: PM is started at port 30536
- :heavy_check_mark: opening new empty project works on local backend
- :heavy_check_mark: opening new project from template works on local backend
- :heavy_check_mark: opening project works on remote (cloud) backend
- :heavy_check_mark: manually running PM at 30535 fails when there is another application running on that port
- :heavy_check_mark: manually running PM at 30535 succeeds when there is no application running on that port
- :heavy_check_mark: manually running PM without explicit port, defaults to 30535
- :heavy_check_mark: `http://localhost:8080/?engine.projectManagerUrl=ws://127.0.0.1:30535` errors when there is no WebSocket server at port 30535
- :heavy_check_mark: `http://localhost:8080/?engine.projectManagerUrl=ws://127.0.0.1:30536` doesn't error, and properly lists projects

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/ProjectManager.scala [hunk 1]
```diff
@@ -287,21 +312,25 @@ object ProjectManager extends ZIOAppDefault with LazyLogging {
...
 
-  private def logServerStartup(): UIO[Unit] =
+  private def logServerStartup(processConfig: MainProcessConfig): UIO[Unit] =
     ZIO.succeed {
       logger.info(
         "Started server at {}:{}, press enter to kill server",
-        config.server.host,
-        config.server.port
+        processConfig.serverHost.getOrElse(config.server.host),
+        processConfig.serverPort.getOrElse(config.server.port)
```
Comment (4e6) [inline lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/ProjectManager.scala]: Please add overrides in the config (instead or in addition to the cli flags) https://github.com/enso-org/enso/blob/59329bd59aeacc4f60ccc4ef323dfaf6629c7710/lib/scala/project-manager/src/main/resources/application.conf#L35-L36

The same way it is done for the `network` https://github.com/enso-org/enso/blob/59329bd59aeacc4f60ccc4ef323dfaf6629c7710/lib/scala/project-manager/src/main/resources/application.conf#L23-L32

i.e the line 
```
   interface = ${?NETWORK_INTERFACE}
```
defines override of the `interface` value by the `NETWORK_INTERFACE` environment variable.

This way you can override config by providing the necessary environment variables without the need of adding extra CLI parameters.

**Comment by somebody1234** [comment]
oh oops

(edit: the changes aren't finished, and the electron app will need to switch to provide envrironment variables instead, so the latest commit does NOT work)

**Comment by somebody1234** [comment]
so it seems like all the CLI changes were unnecessary - not a bad thing though, since it means fewer things could go wrong

**Comment by somebody1234** [comment]
i've done basic re-testing so there shouldn't be any *major* problems

**Review by 4e6 — APPROVED** [review]

File: lib/scala/project-manager/src/main/resources/application.conf [hunk 2]
```diff
@@ -51,7 +54,7 @@ project-manager {
 
   storage {
     projects-root = ${user.home}/enso
-    projects-root=${?PROJECTS_ROOT}
+    projects-root = ${?PROJECTS_ROOT}
```
Comment (JaroslavTulach) [inline lib/scala/project-manager/src/main/resources/application.conf:57]: I cannot claim to be an expert on Scala and its `.conf` file. However I believe that the `${?SERVER_HOST}` definition reads value from environment property, if it is set.

I am surprised we need _so little changes_ in the `project-manager` code!

File: app/ide-desktop/lib/client/src/index.ts [hunk 3]
```diff
@@ -225,9 +230,29 @@ class App {
...
+            const projectManagerUrl = `ws://${this.projectManagerHost}:${this.projectManagerPort}`
+            this.args.groups.engine.options.projectManagerUrl.value = projectManagerUrl
             const backendOpts = this.args.groups.debug.options.verbose.value ? ['-vv'] : []
-            projectManager.spawn(this.args, backendOpts)
+            const backendEnv = Object.assign({}, process.env, {
+                // These are environment variables, and MUST be in CONSTANT_CASE.
+                // eslint-disable-next-line @typescript-eslint/naming-convention
+                SERVER_HOST: this.projectManagerHost,
+                // eslint-disable-next-line @typescript-eslint/naming-convention
+                SERVER_PORT: `${this.projectManagerPort}`,
```
Comment (JaroslavTulach) [inline app/ide-desktop/lib/client/src/index.ts:253]: This code prepares the `SERVER_HOST` and `SERVER_PORT` environment variables for the `project-manager`.

File: app/ide-desktop/lib/client/src/bin/project-manager.ts [hunk 4]
```diff
@@ -30,23 +30,28 @@ export function pathOrPanic(args: config.Args): string {
 }
 
 /** Execute the Project Manager with given arguments. */
-async function exec(args: config.Args, processArgs: string[]) {
+async function exec(args: config.Args, processArgs: string[], env?: NodeJS.ProcessEnv) {
     const binPath = pathOrPanic(args)
-    return await execFile(binPath, processArgs)
+    return await execFile(binPath, processArgs, { env })
```
Comment (JaroslavTulach) [inline app/ide-desktop/lib/client/src/bin/project-manager.ts:35]: Here the environment variables are passed to the `project-manager` process.

**Review by JaroslavTulach — APPROVED** [review]
As far as I can say this is the right approach and I am looking forward to use it.

**Comment by somebody1234** [comment]
@PabloBuchu this should be QAd (and code reviewed) at some point too btw

**Comment by PabloBuchu** [comment]
just tried build and run normally and it is trowing errors

Screenshot 2023-08-22 at 10 03 52
Screenshot 2023-08-22 at 10 04 02


@JaroslavTulach @4e6 could you do a QA of the part with the PM port?

**Comment by somebody1234** [comment]
weird, i thought `MutuallyExclusiveOptions` wouldn't be an issue after being fixed elsewhere... maybe something related to the project manager is staying in the URL?

i did change the logic that decides whether to try loading `index.js.gz` in the fix for `showLogs`, i guess it still always needs to `history.replaceState(null, '', new URL('.', originalUrl))` when opening cloud projects though
https://github.com/enso-org/enso/blob/a2946d2d2740b44314eef1a02ab02cb94762327b/app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/editor.tsx#L122-L139

**Comment by somebody1234** [comment]
pushed a commit that *may* fix it, unable to test due to issues with linux PM bundle

**Comment by 4e6** [comment]
I did the QA. It seems to work on my Linux machine
- [x] Started the project manager on a custom port `env SERVER_PORT=30536 java -jar $JAVA_OPTS project-manager.jar -vv` and then connected to it from the browser `http://localhost:8080/?authentication=false&engine.projectManagerUrl=ws://127.0.0.1:30536`
- [x] Built the IDE distribution with `./run --skip-version-check ide build --skip-wasm-opt` and started it with `./dist/ide/enso-linux-2023.2.1-dev.AppImage -authentication=false --debug.dev-tools` without errors in the console.

**Review by PabloBuchu — APPROVED** [review]

**Pull request merged**